### PR TITLE
Drop root privileges before forking

### PR DESCRIPTION
**Pull request by teto519f**
Implemented a way for lo2s to detect a call with sudo and then drop its privileges before launching any process. Different methods of gaining root aren't taken into account (yet).

This refers to #301 

To Do:

- [x] test functionality in real use case
- [x] ensure compatibility with lo2s -a
- [x] look into doas support (then update manpage)

### CODE REVIEW

**Comment by teto519f** [comment]
With current changes you cannot call on any process, that needs sudo through lo2s. 
For example `lo2s -- sudo apt update` does not work, since lo2s can't call on sudo without root privileges.
`sudo lo2s -- sudo apt update` should be something you're able to do (imo), but this doesn't work when dropping privileges.

TL;DR I would like to add an option, that disables the sudo/doas check. Is this valid?

**Comment by cvonelm** [comment]
> TL;DR I would like to add an option, that disables the sudo/doas check. Is this valid?

I would do it the other way around, enabling this getuid functionality only if some option is given. This feels like something that might lead to weird-to-detect errors if lo2s would just do privilege dropping automatically

Ideally, this option would also accept some username, so that you could specify that you want to switch to that specific user.

You can get the UID/GID of some username by using getpwnam (https://linux.die.net/man/3/getpwnam)

File: man/lo2s.1.pod [hunk 1]
```diff
@@ -154,6 +154,14 @@ determine the trace path instead, including variable substitution.
 Attach to a running process with process ID I<PID> instead of launching
 I<COMMAND>.
 
+=item B<-u>, B<--as-current-user>
+
+Launch I<COMMAND> as the current user. Requires lo2s call with sudo.
+
+=item B<-U>, B<--as-user> I<USERNAME>
+
+Launch I<COMMAND> as the specified user. Implies B<-u> but does not require sudo, but the current user has to have setuid() privileges. Will take priority over B<-u> if specified together.
```
Comment (cvonelm) [inline man/lo2s.1.pod:163]: user I<USERNAME> instead of "the specified user" . I think the part about "implies" in its current implementation confuses more than it helps

File: src/config.cpp [hunk 2]
```diff
@@ -185,6 +185,15 @@ void parse_program_options(int argc, const char** argv)
         .metavar("PID")
         .optional();
 
+    general_options
+        .toggle("as-current-user", "Drop root privileges to current user before launching COMMAND.")
+        .short_name("u");
+
+    general_options.option("as-user", "Specify a user to drop privileges to.")
```
Comment (cvonelm) [inline src/config.cpp:192]: change wording in accordance with man page comment above.

File: src/config.cpp [hunk 3]
```diff
@@ -421,6 +431,18 @@ void parse_program_options(int argc, const char** argv)
         }
     }
 
+    if (arguments.given("as-current-user") && std::getenv("SUDO_UID") == NULL)
+    {
+        Log::error() << "-u was specified but no sudo was used.";
```
Comment (cvonelm) [inline src/config.cpp:436]: used -> detected

File: man/lo2s.1.pod [hunk 4]
```diff
@@ -154,13 +154,13 @@ determine the trace path instead, including variable substitution.
...
-=item B<-u>, B<--as-current-user>
+=item B<-u>, B<--as-pre-sudo-user>
 
-Launch I<COMMAND> as the current user. Requires lo2s call with sudo.
+Launch I<COMMAND> as the user that called on sudo. Requires a lo2s call with sudo.
 
 =item B<-U>, B<--as-user> I<USERNAME>
 
-Launch I<COMMAND> as the specified user. Implies B<-u> but does not require sudo, but the current user has to have setuid() privileges. Will take priority over B<-u> if specified together.
+Launch I<COMMAND> as I<USERNAME>. I<USERNAME> has to have setuid() privileges. Will take priority over B<-u> if specified together.
```
[the remark this answers is not in the document]
Comment (teto519f) [reply man/lo2s.1.pod:163]: Do you think "the caller" or "the calling user" is a good description for this user? Just wanted to ask before I commit again.

[the remark this answers is not in the document]
Comment (cvonelm) [reply man/lo2s.1.pod:163]: "the caller needs to have CAP_SETUID" is the most correct way to write it me thinks

**Review by cvonelm — APPROVED** [review]

**Comment by cvonelm** [comment]
I have tested this now with our main use-case (which is running the IOR benchmark with mpirun, while lo2s runs as root) which seems to run fine, so we can check of testing

File: src/config.cpp [hunk 5]
```diff
@@ -185,6 +185,14 @@ void parse_program_options(int argc, const char** argv)
         .metavar("PID")
         .optional();
 
+    general_options.toggle("as-pre-sudo-user", "Drop root privileges before launching COMMAND.")
```
Comment (bmario) [inline src/config.cpp:188]: The internal config option is called `drop_root`, the description is "Drop root privileges ...", but the thing is called `--as-pre-sudo-user`. I think that is confusing. Don't differ between "internal" and "external" naming of features. I think `--drop-root(-privileges)` would have been a better name.

File: src/config.cpp [hunk 6]
```diff
@@ -185,6 +185,14 @@ void parse_program_options(int argc, const char** argv)
...
+        .short_name("u");
+
+    general_options.option("as-user", "User to drop privileges to before launching COMMAND.")
```
Comment (bmario) [inline src/config.cpp:191]: I think this is easier to understand:
"Launch the COMMAND as the user USERNAME"

File: src/config.cpp [hunk 7]
```diff
@@ -421,6 +430,18 @@ void parse_program_options(int argc, const char** argv)
         }
     }
 
+    if (arguments.given("as-pre-sudo-user") && std::getenv("SUDO_UID") == NULL)
```
Comment (bmario) [inline src/config.cpp:433]: This is not the behavior I would expect.
While `lo2s -a -A` works fine, `lo2s -u -U $USER` complains about `-u was specified but no sudo was detected.`

Or in other words, `-U` should imply `-u`.

Comment (bmario) [reply src/config.cpp:433]: Also, in C++, we don't use `NULL`. Use `nullptr` instead.

File: src/monitor/process_monitor_main.cpp [hunk 8]
```diff
@@ -53,6 +55,72 @@ namespace lo2s
...
+            Log::error() << "No user found with username '" + config().user + ".";
+            throw_errno();
+        }
+
+        original_uid = orig_user->pw_uid;
+        original_gid = orig_user->pw_gid;
+    }
+    else
+    {
+        // sudo is garanteed in this case, see check in config.hpp
```
Comment (bmario) [inline src/monitor/process_monitor_main.cpp:78]: Don't only put your assumptions into comments. Put them in `assert()`s.

File: src/monitor/process_monitor_main.cpp [hunk 9]
```diff
@@ -53,6 +55,72 @@ namespace lo2s
...
+        try
+        {
+            original_uid = std::stoi(std::getenv("SUDO_UID"));
+            original_gid = std::stoi(std::getenv("SUDO_GID"));
+        }
+        catch (std::invalid_argument const& e)
+        {
+            Log::error() << "Cannot parse SUDO_UID/SUDO_GID into int.";
```
Comment (bmario) [inline src/monitor/process_monitor_main.cpp:86]: This will be a user-facing message. Users don't know what an `int` is or that `SUDO_UID/SUDO_GID` are environment variables.

`Cannot parse the environment variables SUDO_UID and/or SUDO_GID.`

File: src/monitor/process_monitor_main.cpp [hunk 10]
```diff
@@ -53,6 +55,72 @@ namespace lo2s
...
+            throw_errno();
```
Comment (bmario) [inline src/monitor/process_monitor_main.cpp:87]: `std::stoi` raises an exception, but does not set `errno`. `throw_errno()` would need that.

Replace with `throw;`, which rethrows the current exception.

File: src/monitor/process_monitor_main.cpp [hunk 11]
```diff
@@ -53,6 +55,72 @@ namespace lo2s
...
+        }
+        catch (std::out_of_range const& e)
```
Comment (bmario) [inline src/monitor/process_monitor_main.cpp:89]: I don't think the distinction between both exceptions types adds any value here. I'd rather have one "parsing failed" message.

File: src/monitor/process_monitor_main.cpp [hunk 12]
```diff
@@ -53,6 +55,72 @@ namespace lo2s
...
+        {
+            Log::error() << "SUDO_UID/SUDO_GID out of range.";
+            throw_errno();
```
Comment (bmario) [inline src/monitor/process_monitor_main.cpp:92]: `throw_errno()` doesn't work here, but remove it completely.

File: src/monitor/process_monitor_main.cpp (hunk 11 above)
Comment (teto519f) [reply src/monitor/process_monitor_main.cpp:89]: Would you remove the second catch statement and catch general exceptions or duplicate the error handling for both types? 
I'd go for the second option, but redundant code is wack and a function would be overkill.

Comment (bmario) [reply src/monitor/process_monitor_main.cpp:89]: I'd go with one handler `catch(const std::exception& e)`. As users typically do not set these environment variables themselves, I don't think any other message than "failed to parse ..."  is necessary, but rather confusing.
